### DISCUSSION TITLE
Insert statement for a mapped(renamed) column name

### DISCUSSION THREAD

**Discussion by claria** [sqlalchemy/sqlalchemy#9970] [Usage Questions]
Hi,

i have a ORM table design where the ORM attribute is named differently than the sql column

```python
class Device(Base):
    id = Column(Integer, primary_key=True)
    comment = Column('remarks', String(100))
```

I often query with select statements like this

`stmt = sa.select(Device.id, Device.comment)`

which works nicely. 

Sometimes, i would like to insert a whole bunch of values into the table and always did it (on other tables) like this:

```
stmt = sa.insert(Device)
session.execute(stmt, [{'id': 1, 'comment': 'asdf'},{'id':2, 'comment': '1234'}])
```

However, in this case this leads to `CompileError: Unconsumed column names: comment` since it cannot map the 'comment' attribute name to the sql column. 

What i do not understand ist that 

```
print(sa.insert(Device).values(comment='asdf'))
CompileError: Unconsumed column names: comment
```

does not work while

```
print(sa.update(Device).values(comment='asdf'))
UPDATE device SET remarks=:remarks
```
generates the correct update statement.

Is there a simple way to either get the mapping between ORM attribute name and sql column name to replace the keys in the dictionaries or is there another way to supply the values and tell sqlalchemy how to perform the mapping?

Best regards and thanks for your help!

sqlalchemy 1.4.48

**Comment by CaselIT** [comment]
Hi,

The fact that insert does not work while update does seems like a bug. Thanks for reporting. Have you tried by any chance 2.0?

**Comment by claria** [comment]
Hi,

just setup this small example in sqlalchemy 2.0. There everything works fine.

```
from typing import List
from typing import Optional
from sqlalchemy import ForeignKey
from sqlalchemy import String
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
import sqlalchemy as sa

class Base(DeclarativeBase):
    pass

class User(Base):
    __tablename__ = "user_account"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column("old_name", String(30))


print(sa.update(User).values(name='asdf'))
print(sa.insert(User).values(name='asdf'))
```

**Reply by CaselIT** [reply]
Thanks for checking

**Comment by claria** [comment]
Is this something which still would be fixed in 1.4.* ? Then i would open an issue with a real bug report.

**Reply by CaselIT** [reply]
This may be a limitation in the 1.4 series, since not all dml statements are orm enabled. cc @zzzeek

**Reply by zzzeek** [reply]
ORM enabled INSERT was a major feature add in 2.0: https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#orm-enabled-insert-upsert-update-and-delete-statements-with-orm-returning

not a bug in 1.4 as this is expected behavior for 1.4.   the change is backwards incompatible as it changes the naming scheme used to evaluate INSERT statements.

**Comment by claria** [comment]
Here is the adapted example for sqlalchemy 1.4 where the bug occurs:

```
from typing import List
from typing import Optional
from sqlalchemy import ForeignKey
from sqlalchemy import String
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import relationship
import sqlalchemy as sa

from sqlalchemy.orm import declarative_base
Base = declarative_base()

class User(Base):
    __tablename__ = "user_account"

    id: Mapped[int] = sa.Column(primary_key=True)
    name: Mapped[str] = sa.Column("old_name", String(30))


print(sa.update(User).values(name='asdf'))
print(sa.insert(User).values(name='asdf'))
```

**Reply by zzzeek** [reply]
this is specifically #7864
